controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -331,18 +334,71 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
+	if err != nil {
+		return err
+	}
+	requireUpgrade := internal.FilterMachines(
+		ownedMachines,
+		internal.HasOutdatedConfiguration(hash.Compute(&kcp.Spec)),
+		internal.OlderThan(kcp.Spec.UpgradeAfter),
+	)
+
+	if err := r.managementCluster.TargetClusterEtcdIsHealthy(ctx, clusterKey(cluster), kcp.Name); err != nil {
+		logger.Error(err, "Waiting for etcd cluster to pass health check")
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ETCDClusterUnhealthy", "Waiting for etcd cluster to be healthy to continue upgrade: %v", err)
+		return &capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second}
+	}
+	if err := r.managementCluster.TargetClusterControlPlaneIsHealthy(ctx, clusterKey(cluster), kcp.Name); err != nil {
+		logger.Error(err, "Waiting for control plane to pass health check")
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ControlPlaneUnhealthy", "Waiting for control plane to be healthy to continue upgrade: %v", err)
+		return &capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second}
+	}
+
+	// If there is not already a Machine that is marked for upgrade, find one and mark it
+	selectedForUpgrade := internal.FilterMachines(requireUpgrade, internal.SelectedForUpgrade())
+	if len(selectedForUpgrade) == 0 {
+		sort.Slice(requireUpgrade, func(i, j int) bool {
+			return requireUpgrade[i].CreationTimestamp.Before(&requireUpgrade[j].CreationTimestamp)
+		})
+
+		oldest := &requireUpgrade[0]
+		patchHelper, err := patch.NewHelper(oldest, r.Client)
+		if err != nil {
+			return err
+		}
+
+		oldest.Labels[controlplanev1.SelectedForUpgradeLabel] = ""
+
+		if err := patchHelper.Patch(ctx, oldest); err != nil {
+			return err
+		}
+		selectedForUpgrade = append(selectedForUpgrade, *oldest)
+	}
+
+	// Determine if we need to add an additional control plane Machine or clean up an already replaced one
+	switch {
+	case len(ownedMachines) == int(*kcp.Spec.Replicas):
+		// return here to avoid blocking while waiting for the new control plane Machine to come up
+		return r.scaleUpControlPlane(ctx, cluster, kcp, 1)
+	default:
+		sort.Slice(selectedForUpgrade, func(i, j int) bool {
+			return selectedForUpgrade[i].CreationTimestamp.Before(&selectedForUpgrade[j].CreationTimestamp)
+		})
+		machineToDelete := selectedForUpgrade[0]
+		if err := r.managementCluster.RemoveEtcdMemberForMachine(ctx, clusterKey(cluster), machineToDelete); err != nil {
+			return err
+		}
+		// TODO: Update the kubeadm configmap
+		if err := r.Client.Delete(ctx, &machineToDelete); err != nil && !apierrors.IsNotFound(err) {
+			logger.WithValues("machine", machineToDelete.Name).Error(err, "Failed to delete marked Machine")
+			r.recorder.Eventf(kcp, corev1.EventTypeWarning, "FailedDeleteMachine", "Failed to delete Machine %s marked for deletion: %v", machineToDelete.Name, err)
+			return &capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second}
```

Comment:
why does this RequeueAfter instead of returning an error?

Reply:
Mainly because I was being a bit lazy and didn't want to have to distinguish retryable and non-retryable errors returned from this function (other than checking if they are a RequeueAfterError) :joy:  I'll clean it up.